tsl/test/sql/recompress_chunk_segmentwise.sql

```diff
@@ -0,0 +1,134 @@
+
+select show_chunks as chunk_to_compress_2 from show_chunks('mytab_twoseg') limit 1 \gset
+
+select compress_chunk(:'chunk_to_compress_2');
+
+insert into mytab_twoseg values ('2023-01-01 19:56:20.048355+02'::timestamptz, 2, NULL, 2);
+
+select * from :chunk_to_compress_2;
+
+SELECT compressed_chunk_schema || '.' || compressed_chunk_name as compressed_chunk_name_2
+from compressed_chunk_info_view where hypertable_name = 'mytab_twoseg' \gset
+
+select ctid, * from :compressed_chunk_name_2;
+
+select _timescaledb_internal.recompress_chunk_segmentwise(:'chunk_to_compress_2');
+
+-- verify that metadata count looks good
+select ctid, * from :compressed_chunk_name_2;
+
+-- verify that initial data is returned as expected
+select * from :chunk_to_compress_2;
+
+----------------- more than one batch per segment ----------------------
+-- test that metadata sequence number is correct
+create table mytab2(time timestamptz not null, a int, b int, c int);
+
+select create_hypertable('mytab2', 'time', chunk_time_interval => interval '1 week');
+
+insert into mytab2 (time, a, c) select t,s,s from
+generate_series('2023-01-01 00:00:00+00'::timestamptz, '2023-01-01 00:00:00+00'::timestamptz + interval '1 day', interval '30 sec') t cross join generate_series(0,2, 1) s;
+
+alter table mytab2 set (timescaledb.compress, timescaledb.compress_segmentby = 'a, c');
+
+select compress_chunk(c) from show_chunks('mytab2') c;
+
+SELECT compressed_chunk_schema || '.' || compressed_chunk_name as compressed_chunk_name_2
+from compressed_chunk_info_view where hypertable_name = 'mytab2'
+and compressed_chunk_name is not null limit 1 \gset
+
+insert into mytab2 values ('2023-01-01 00:00:02+00'::timestamptz, 0, NULL, 0); -- goes into the uncompressed chunk
+
+select show_chunks('mytab2') as chunk_to_compress_2 \gset
+
+select ctid, * from :compressed_chunk_name_2;
+
+select _timescaledb_internal.recompress_chunk_segmentwise(:'chunk_to_compress_2');
+
+select ctid, * from :compressed_chunk_name_2;
+
+-- failing test from compression_ddl
+CREATE TABLE test_defaults(time timestamptz NOT NULL, device_id int);
+SELECT create_hypertable('test_defaults','time');
+
+ALTER TABLE test_defaults SET (timescaledb.compress,timescaledb.compress_segmentby='device_id');
+
+-- create 2 chunks
+INSERT INTO test_defaults SELECT '2000-01-01', 1;
+INSERT INTO test_defaults SELECT '2001-01-01', 1;
+
+SELECT compress_chunk(show_chunks) AS "compressed_chunk" FROM show_chunks('test_defaults') ORDER BY show_chunks::text LIMIT 1 \gset
+
+SELECT * FROM test_defaults ORDER BY 1;
+
+ALTER TABLE test_defaults ADD COLUMN c1 int;
+ALTER TABLE test_defaults ADD COLUMN c2 int NOT NULL DEFAULT 42;
+SELECT * FROM test_defaults ORDER BY 1,2;
+
+INSERT INTO test_defaults SELECT '2000-01-01', 2;
+SELECT * FROM test_defaults ORDER BY 1,2;
+
+call recompress_chunk(:'compressed_chunk');
+SELECT * FROM test_defaults ORDER BY 1,2;
```

Comment:
Do you have a test to check which strategy is chosen by `recompress_chunk`?  

One idea is adding a `DEBUG1` log message to both strategies and checking the chosen strategy in the regression tests.

Reply:
Added, as suggested

Reply:
I also need to find a way to remove/suppress some additional debug prints so I unresolved the conversation until then

Reply:
in order to test this, I removed the debug prints and check whether the compressed chunk after recompression is the same or not because this differs depending on the strategy. So marking conversation as resolved.